operator-common/src/main/java/io/strimzi/operator/common/metrics/MetricsHolder.java

```diff
@@ -5,31 +5,79 @@
 package io.strimzi.operator.common.metrics;
 
 import io.micrometer.core.instrument.Counter;
-import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Meter;
 import io.micrometer.core.instrument.Tags;
 import io.micrometer.core.instrument.Timer;
 import io.strimzi.operator.common.MetricsProvider;
-import io.strimzi.operator.common.config.ConfigParameter;
 import io.strimzi.operator.common.model.Labels;
 
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
  * Abstract base class holding common metrics used by operators and controllers.
  * Subclasses can add more specialized metrics.
  */
 public abstract class MetricsHolder {
-    protected static final String METRICS_PREFIX = "strimzi.";
+    /**
+     * Prefix used for metrics provided by Strimzi operators
+     */
+    public static final String METRICS_PREFIX = "strimzi.";
+    /**
+     * Metric name for number of reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS = METRICS_PREFIX + "reconciliations";
+    /**
+     * Metric name for number of periodic reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS_PERIODICAL = METRICS_PREFIX + "reconciliations.periodical";
+    /**
+     * Metric name for number of failed reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS_FAILED = METRICS_PREFIX + "reconciliations.failed";
+    /**
+     * Metric name for number of successful reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS_SUCCESSFUL = METRICS_PREFIX + "reconciliations.successful";
+    /**
+     * Metric name for duration of reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS_DURATION = METRICS_PREFIX + "reconciliations.duration";
+    /**
+     * Metric name for number of locked reconciliations.
+     */
+    public static final String METRICS_RECONCILIATIONS_LOCKED = METRICS_PREFIX + "reconciliations.locked";
+    /**
+     * Metric name for number of resources managed by the operator.
+     */
+    public static final String METRICS_RESOURCES = METRICS_PREFIX + "resources";
+    /**
+     * Metric name for number of paused resources.
+     */
+    public static final String METRICS_RESOURCES_PAUSED = METRICS_PREFIX + "resources.paused";
+    /**
+     * Metric name for server certificate expiration timestamp in ms.
+     */
+    public static final String METRICS_SERVER_CERTIFICATE_EXPIRATION_MS = METRICS_PREFIX + "server.certificate.expiration.ms";
+    /**
+     * Metric name for client certificate expiration timestamp in ms.
+     */
+    public static final String METRICS_CLIENT_CERTIFICATE_EXPIRATION_MS = METRICS_PREFIX + "client.certificate.expiration.ms";
```

Comment:
One day, we might be asked about metrics for the actual server and user certificates (I'm pretty sure the user certs will come soon after you finish this 😉). So I wonder ... should we have a single metric named `certificate.expiration.ms` and separate by tags? If not, I would call these `cluster.ca.certificate.expiration` and `clients.ca.certificate.expiration` or something like that to make them specific for what they are.

---

I also wonder how extensible this is for using multiple client CAs which is another requirement that might come. But that is more of a thought, we might not be able to design for it today.

Reply:
Yeah, if it cannot be done with tags, I guess we can live with it. But it should then have the improved names.

Reply:
Sure

Reply:
Found a way to do it, I've extended `metrics()` function in `MetricsHolder` to include optional list of additional Tags `Tag... optionalTags` which whatever metric then can be extended to.

Furthermore, to eliminate the the string split, we spoke about, which I found quite unreadable, when I did it. I've introduced the concept of `MetricKey` used to index in the maps, where you can then again retrieve the different elements.